pallets/rate_provider/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+#![allow(
+    clippy::unused_unit,
+    unused_qualifications,
+    missing_debug_implementations
+)]
+#![cfg_attr(not(feature = "std"), no_std)]
+use codec::{Decode, Encode};
+pub use pallet::*;
+pub use vln_primitives::{PaymentMethod, RateProvider};
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Encode, Decode, Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Rates<CurrencyId, BaseCurrencyId> {
+    pub from: CurrencyId,
+    pub to: BaseCurrencyId,
+    pub method: PaymentMethod,
+}
+
+pub mod primitives {
+    // NOTE We should be able to make our module generic over any kind of `PerThing`
+    pub type LpRatePremium = sp_runtime::Permill;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+    use crate::{primitives::LpRatePremium, PaymentMethod, RateProvider, Rates};
+    use frame_support::{
+        dispatch::DispatchResultWithPostInfo, pallet_prelude::*, traits::Contains,
+    };
+    use frame_system::pallet_prelude::*;
+    use orml_traits::DataProvider;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// Because this pallet emits events, it depends on the runtime's definition of an event.
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        /// Type of assets that the LP can set rates for
+        type CurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
+        /// Type of the base currency
+        type BaseCurrencyId: Parameter + Member + Copy + MaybeSerializeDeserialize + Ord;
+        /// type of Oracle Provider
+        type OracleProvider: DataProvider<Self::CurrencyId, Self::OracleValue>;
+        /// type of Oracle Value
+        type OracleValue: Parameter + Member + Ord;
+        /// Whitelist of LPs allowed to participate, this will eventually be removed and
+        /// anyone should be able to publish rates
+        type Whitelist: Contains<Self::AccountId>;
+    }
+
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    #[pallet::storage]
+    #[pallet::getter(fn rates)]
+    // Rates provided by an LP, the actual rate is not stored here
```

Comment:
In the context of this pallet let's just call them "providers" as the liquidity part is more of a Valiu concept that might not apply for other projects